sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java

```diff
@@ -116,17 +121,25 @@ public boolean requiresDeduping() {
      * <ul>
      * <li>This finalize method may be called from any thread, concurrently with calls to
      * the {@link UnboundedReader} it was created from.
+     * <li>This method may be called on a {@link CheckpointMark} that has been decoded by the
+     * {@link Coder} returned by {@link #getCheckpointMarkCoder()}.
      * <li>Checkpoints will not necessarily be finalized as soon as they are created.
-     * A checkpoint may be taken while a previous checkpoint from the same
+     * <li>A checkpoint may be taken while a previous checkpoint from the same
      * {@link UnboundedReader} has not yet be finalized.
-     * <li>In the absence of failures, all checkpoints will be finalized and they will be
-     * finalized in the same order they were taken from the {@link UnboundedReader}.
      * <li>It is possible for a checkpoint to be taken but this method never called. This method
      * will never be called if the checkpoint could not be committed, and other failures may cause
      * this method to never be called.
+     * <li>A {@link PipelineRunner} may finalize a checkpoint more than once.
+     * <li>A {@link PipelineRunner} may finalize a checkpoint that has been decoded with the result
+     * of {@link UnboundedSource#getCheckpointMarkCoder()}.
+     * <li>In the absence of failures, all checkpoints will be finalized and they will be
+     * finalized in the same order they were taken from the {@link UnboundedReader}.
      * <li>It is not safe to assume the {@link UnboundedReader} from which this checkpoint was
      * created still exists at the time this method is called.
      * </ul>
+     *
+     * <p>As a result of the above requirements, {@link #finalizeCheckpoint()} must be idempotent
+     * and contain any state required to perform finalization.
```

Comment:
Given the situation with pubsub - which may be repeated, who knows - we should mention "it is permitted for a checkpoint to perform less work or no work at all during finalization, especially if it has been serialized then deserialized"

Of course, given that fact, there might not be much value to idempotency. The alternative is to embrace the weirdness and say "A runner is only permitted to call `finalizeCheckpoint` for a checkpoint that was just taken from a reader. A deserialized checkpoint must never be finalized."

It does sound like if finalization is only _useful_ in the above case, then no runner is going to be finalizing after a deserialization. But on the other hand, if it releases resources it seems good to finalize ASAP after deserialization. So I am generally in favor of loosening the spec, but first making sure that is sensible.